plugins/chain_plugin/trx_finality_status_processing.cpp

```diff
@@ -1,12 +1,215 @@
+   : _my(new trx_finality_status_processing_impl(max_storage, success_duration, failure_duration))
    {
    }
 
+   trx_finality_status_processing::~trx_finality_status_processing() {
+   }
+
+   void trx_finality_status_processing::signal_applied_transactions( const chain::signals_processor::trx_deque& trxs, const chain::block_state_ptr& bsp ) {
+      _my->signal_applied_transactions(trxs, bsp);
+   }
+
+   void trx_finality_status_processing::signal_irreversible_block( const chain::block_state_ptr& bsp ) {
+      _my->_irr_block_id = bsp->id;
+   }
+
+   void trx_finality_status_processing::signal_block_start( uint32_t block_num ) {
+   }
+
+   void trx_finality_status_processing_impl::signal_applied_transactions( const chain::signals_processor::trx_deque& trxs, const chain::block_state_ptr& bsp ) {
+      const fc::time_point now = fc::time_point::now();
+      // use the head block num if we are in a block, otherwise don't provide block number for speculative blocks
+      const auto block_id = bsp ? bsp->id : chain::block_id_type{};
+      bool modified = !trxs.empty();
+      if (bsp) {
+         _head_block_id = block_id;
+         if (chain::block_header::num_from_id(_head_block_id) <= _last_proc_block_num) {
+            handle_rollback();
+            modified = true;
+         }
+
+         // clean up all status expired transactions, to free up storage
+         if (status_expiry_of_trxs(now)) {
+            modified = true;
+         }
+      }
+
+      for (const auto& trx_tuple : trxs) {
+         ensure_storage(now);
+         const auto& trx_id = std::get<0>(trx_tuple)->id;
+         auto iter = _storage.find(trx_id);
+         if (iter != _storage.index().cend()) {
+            _storage.modify( iter, [&block_id]( finality_status_object& obj ) {
+               obj.block_id = block_id;
+            } );
+         }
+         else {
+            _storage.insert(finality_status_object{.trx_id = trx_id, .trx_expiry = std::get<1>(trx_tuple)->expiration(), .received = now, .block_id = block_id});
+         }
+      }
+      if (modified) {
+         const auto& indx = _storage.index().get<by_block_num>();
+         auto iter = indx.cbegin();
+         if (iter != indx.cend()) {
+            _last_tracked_block_id = iter->block_id;
+         }
+      }
+
+      if (bsp) {
+         _last_proc_block_num = chain::block_header::num_from_id(_head_block_id);
+      }
+   }
+
+   void trx_finality_status_processing_impl::handle_rollback() {
+      const auto& indx = _storage.index().get<by_block_num>();
+      std::deque<chain::transaction_id_type> trxs;
+      auto iter = indx.lower_bound(_head_block_id);
+      std::transform(iter, indx.end(), trxs.end(), []( const finality_status_object& obj ) { return obj.trx_id; });
+      for (const auto& trx_id : trxs) {
+         auto trx_iter = _storage.find(trx_id);
+         _storage.modify( trx_iter, []( finality_status_object& obj ) {
+            obj.block_id = chain::block_id_type{};
```

Comment:
Do you want to zero that out? Isn't possible useful info what block it was in even if forked out?

Reply:
I think you are correct, but will need to change around my indexes to use methods, so not-valid block can also check forked_out == true || block_id == block_id_type{}

Reply:
This is the only change I haven't tried to address yet